zerostash/src/commands/mount.rs

```diff
@@ -0,0 +1,381 @@
+            .expect("Could not send signal on channel.")
+    }
+    fn getattr(&self, _req: RequestInfo, path: &Path, _fh: Option<u64>) -> ResultEntry {
+        debug!("gettattr = {:?}", path);
+
+        let real_path = strip_path(path);
+
+        if self.stash.index().directories.contains(&path.to_path_buf()) {
+            Ok((TTL, DIR_ATTR))
+        } else {
+            let path_string = real_path.to_str().unwrap();
+            match self.stash.index().files.get(path_string) {
+                Some(metadata) => {
+                    let fuse = file_to_fuse(&metadata, self.commit_timestamp);
+                    Ok((TTL, fuse))
+                }
+                None => Err(libc::ENOENT),
+            }
+        }
+    }
+
+    fn opendir(&self, _req: RequestInfo, _path: &Path, _flags: u32) -> ResultOpen {
+        debug!("opendir");
+        Ok((0, 0))
+    }
+
+    fn readdir(&self, _req: RequestInfo, path: &Path, _fh: u64) -> ResultReaddir {
+        debug!("readdir: {:?}", path);
+        let entries = self.stash.index().directories.get(path).unwrap_or_default();
+        let entries = entries.lock().unwrap();
+        let transformed_entries = transform(entries.to_vec());
+        Ok(transformed_entries)
+    }
+    fn open(&self, _req: RequestInfo, path: &Path, _flags: u32) -> ResultOpen {
+        debug!("open: {:?}", path);
+        let real_path = strip_path(path);
+        let mut stack = self.stack.lock().unwrap();
+        stack.insert(real_path.to_path_buf(), ChunkDataStack::default());
+        Ok((0, 0))
+    }
+    fn read(
+        &self,
+        _req: RequestInfo,
+        path: &Path,
+        _fh: u64,
+        offset: u64,
+        size: u32,
+        callback: impl FnOnce(ResultSlice<'_>) -> CallbackResult,
+    ) -> CallbackResult {
+        debug!("read: {:?} {:#x} @ {:#x}", path, size, offset);
+
+        let real_path = strip_path(path);
+        let path_string = real_path.to_str().unwrap();
+        let metadata = self.stash.index().files.get(path_string).unwrap();
+        let file_size = metadata.size as usize;
+        let offset = offset as usize;
+
+        if offset > file_size {
+            return callback(Err(libc::EINVAL));
+        }
+
+        let size = size as usize;
+        let mut chunks = metadata.chunks.clone();
+        chunks.sort_by(|(a, _), (b, _)| a.cmp(b));
+        let mut objectreader = self.stash.storage_reader().unwrap();
+
+        {
+            let mut stack = self.stack.lock().unwrap();
+            let stack = stack
+                .entry(real_path.to_path_buf())
+                .or_insert_with(ChunkDataStack::default);
+
+            if stack.last_read_offset == offset {
+                let end = size.min(file_size - offset);
+                if stack.stack.len() < end {
+                    loop {
+                        if stack
+                            .add_chunks(&chunks, file_size, &mut objectreader)
+                            .is_err()
```

Comment:
Since you are cloning and re-sorting `chunks`, you could abstract over an iterator of it instead of pure indexes. So, eg. you could have a type such as:

```rust
struct ChunkStack<T> {
  chunks: T
}

impl<T: Iterator<Target=(u64, ChunkPointer)> ChunkStack<T> {
 fn new(chunks: T) -> Self {
    Self { chunks }
 }
 fn read_next(&mut self, size: u64) -> Vec<u8> {
    let buffer = vec![];
    while !is_full(buffer, size) {
    let next_chunk = self.chunks.next();
    ...
    }
    buffer
 } 
}
```

This would make this kind of code a bit easier to read and reason about. Please do bear in mind that I completely just made these functions up, so your implementation will look different.

What you have is a good start, but the control flow is a bit complicated for my taste, and I think this would help make it more legible at least in this spot.